File: ruby/LinearInsertionSort.rb

# file_name = Rational.rb
require 'F:\ruby\383\A2\Rational.rb'

AVGTESTINGTIMES = 10

def insertion_sort(array)
    for i in 1...(array.length) 
        j = i 
        while j > 0 
            if array[j-1] > array[j]
                temp = array[j]
                array[j] = array[j-1]
                array[j-1] = temp
            else
                break
            end
            j = j - 1 
        end
    end
    return array
end

def insertion_sort_rationalNum(array)
    for i in 1...(array.length) 
        j = i 
        while j > 0 
            if (array[j-1] <=> array[j]) == 1
                temp = array[j]
                array[j] = array[j-1]
                array[j-1] = temp
            else
                break
            end
            j = j - 1 
        end
    end
    return array
end


### Generating data ######
def GeneratingInt(size)
    arr =[]
    size.times do |i|
        arr.push(rand(-2**15..2**15))
    end
    return arr
end

def GeneratingString(size)
    arr =[]
    size.times do |i|
        arr.push(randomString(5))
    end
    return arr
end

def GeneratingRationalNum(size)
    arr = []
    size.times do |i|
        numerator = 0
        denominator = 0
        while denominator == 0 do
            numerator = rand(-2**15..2**15)
            denominator = rand(-2**15..2**15)
        end 
        arr.push(MyRational.new(numerator,denominator))
    end
    return arr
end

def randomString( len )
    chars = ("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a
    randomString = ""
    1.upto(len) { |i| randomString << chars[rand(chars.size-1)] }
    return randomString
end


### calculate cost 
def IntCalCost (n)
    arr = GeneratingInt(n)
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    insertion_sort(arr)
    t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    delta = t2 - t1
    delta_in_microseconds = delta * 1000000
    print "size = #{n}, cost = #{delta_in_microseconds} microseconds, "
end

def StringCalCost (n)
    arr = GeneratingString(n)
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    insertion_sort(arr)
    t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    delta = t2 - t1
    delta_in_microseconds = delta * 1000000
    print "size = #{n}, cost = #{delta_in_microseconds} microseconds, "
end

def RationalNumCalCost (n)
    arr = GeneratingRationalNum(n)
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    insertion_sort_rationalNum(arr)
    t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    delta = t2 - t1
    delta_in_microseconds = delta * 1000000
    print "size = #{n}, cost = #{delta_in_microseconds} microseconds, "
end


## calulate avg 
def IntAvgCost(n)
    sum = 0
    AVGTESTINGTIMES.times do |i|
        arr = GeneratingInt(n)
        t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        insertion_sort(arr)
        t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        sum += t2-t1
    end
    avg = sum/AVGTESTINGTIMES
    avg_microseconds = avg * 1000000
    puts "Sum:#{sum} seconds, Avg:#{avg_microseconds} microseconds"
end

def StringAvgCost(n)
    sum = 0
    AVGTESTINGTIMES.times do |i|
        arr = GeneratingString(n)
        t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        insertion_sort(arr)
        t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        sum += t2-t1
    end
    avg = sum/AVGTESTINGTIMES
    avg_microseconds = avg * 1000000
    puts "Sum:#{sum} seconds, Avg:#{avg_microseconds} microseconds"
end  

def RationalNumAvgCost(n)
    sum = 0
    AVGTESTINGTIMES.times do |i|
        arr = GeneratingRationalNum(n)
        t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        insertion_sort_rationalNum(arr)
        t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        sum += t2-t1
    end
    avg = sum/AVGTESTINGTIMES
    avg_microseconds = avg * 1000000
    puts "Sum:#{sum} seconds, Avg:#{avg_microseconds} microseconds"
end  

## Testing part 
def IntTest()
    puts "--- Int testing ---"
    1000.step(10000,1000) do |i|
        IntCalCost(i)
		IntAvgCost(i)
    end
end

def StringTest()
    puts "--- String testing ---"
    1000.step(10000,1000) do |i|
        StringCalCost(i)
		StringAvgCost(i)
    end
end

def RationalNumTest()
    puts "--- RationalNum testing ---"
    1000.step(10000,1000) do |i|
        RationalNumCalCost(i)
        RationalNumAvgCost(i)
    end
end

### main code ###
# list_int = [7,8,5,1,99,2221,33,0,1]
# list_strings = ["abcgr", "kaeee", "hefpp", "asdgr", "abdgr"]
# list_rationalNum = [MyRational.new(1,2), 
#                     MyRational.new(5,3),  
#                     MyRational.new(5,7), 
#                     MyRational.new(15,7), 
#                     MyRational.new(55,7), 
#                     MyRational.new(5,77), 
#                     MyRational.new(1,99)]

## insertion sort test 
# print list_int
# puts 
# print list_strings
# puts 
# print insertion_sort(list_int)
# puts
# print insertion_sort(list_strings)
# puts
# print list_rationalNum
# puts
# print insertion_sort(list_rationalNum)

# test 
IntTest()
StringTest()
RationalNumTest()

File: go/main/LinearInsertionSort.go
package main

import (
	"bytes"
	"fmt"
	"go_code/A1/RationalType"
	"math"
	"math/rand"
	"time"
)

// This var is for AVG testing. it controls the AVG time spending accuracy
const AVGTESTINGTIMES = 10
const CHAR = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

// Insertion Sort part 1
func InsertionSortInt(items []int) []int {
	var n = len(items)
	for i := 1; i < n; i++ {
		j := i
		for j > 0 {
			if items[j-1] > items[j] {
				items[j-1], items[j] = items[j], items[j-1]
			}
			j = j - 1
		}
	}
	return items
}

func InsertionSortString(items []string) []string {
	var n = len(items)
	for i := 1; i < n; i++ {
		j := i
		for j > 0 {
			if items[j-1] > items[j] {
				items[j-1], items[j] = items[j], items[j-1]
			}
			j = j - 1
		}
	}
	return items
}

func InsertionSortRationalNum(items []RationalType.Rationalizer) []RationalType.Rationalizer {
	var n = len(items)
	for i := 1; i < n; i++ {
		j := i
		for j > 0 {
			if items[j].LessThan(items[j-1]) {
				items[j-1], items[j] = items[j], items[j-1]
			}
			j = j - 1
		}
	}
	return items
}

// GenerateSlice part 2
// random seed range:  	-2^15 to 2^15 -1
func IntGenerateSlice(size int) []int {
	slice := make([]int, size, size)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < size; i++ {
		slice[i] = (rand.Intn(int(math.Pow(2, 15))) - rand.Intn(int(math.Pow(2, 15))))
	}
	return slice
}

func StringGenerateSlice(size int) []string {
	slice := make([]string, size, size)
	rand.NewSource(time.Now().UnixNano())
	for k := 0; k < size; k++ {
		var s bytes.Buffer
		for i := 0; i < 5; i++ {
			s.WriteByte(CHAR[rand.Int63()%int64(len(CHAR))])
		}
		slice[k] = s.String()
	}
	return slice
}

//range from -256 to 255
func RationalNumGenerateSlice(size int) []RationalType.Rationalizer {
	slice := []RationalType.Rationalizer{}
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < size; i++ {
		Num := (rand.Intn(int(math.Pow(2, 15))) - rand.Intn(int(math.Pow(2, 15))))
		Denom := (rand.Intn(int(math.Pow(2, 15))) - rand.Intn(int(math.Pow(2, 15))))
		r, err := RationalType.MakeRational(Num, Denom)
		if err == nil {
			slice = append(slice, r)
		} else {
			i--
		}
	}
	return slice
}

// CalCost part 3
func IntCalCost(N int) {
	list := IntGenerateSlice(N)
	start := time.Now()
	InsertionSortInt(list)
	cost := time.Since(start)
	fmt.Printf("N = %v, cost = %v ", N, cost.Microseconds())
}

func StringCalCost(N int) {
	list := StringGenerateSlice(N)
	start := time.Now()
	InsertionSortString(list)
	cost := time.Since(start)
	fmt.Printf("N = %v, cost = %v ", N, cost.Microseconds())
}

func RationalNumCalCost(N int) {
	list := RationalNumGenerateSlice(N)
	start := time.Now()
	InsertionSortRationalNum(list)
	cost := time.Since(start)
	fmt.Printf("N = %v, cost = %v ", N, cost.Microseconds())
}

// calulate avg part 4
func IntAvgCost(N int) {
	var sum time.Duration
	for i := 0; i < AVGTESTINGTIMES; i++ {
		list := IntGenerateSlice(N)
		start := time.Now()
		InsertionSortInt(list)
		cost := time.Since(start)
		// fmt.Printf(" Sum: %v, cost: %v \n", sum, cost)
		sum += cost
	}
	avg := sum / AVGTESTINGTIMES
	fmt.Printf(" Sum: %v, Avg: %v \n", sum.Microseconds(), avg.Microseconds())
}

func StringAvgCost(N int) {
	var sum time.Duration
	for i := 0; i < AVGTESTINGTIMES; i++ {
		list := StringGenerateSlice(N)
		start := time.Now()
		InsertionSortString(list)
		cost := time.Since(start)
		// fmt.Printf(" Sum: %v, cost: %v \n", sum, cost)
		sum += cost
	}
	avg := sum / AVGTESTINGTIMES
	fmt.Printf(" Sum: %v, Avg: %v \n", sum, avg.Microseconds())
}

func RationalNumAvgCost(N int) {
	var sum time.Duration
	for i := 0; i < AVGTESTINGTIMES; i++ {
		list := RationalNumGenerateSlice(N)
		start := time.Now()
		InsertionSortRationalNum(list)
		cost := time.Since(start)
		// fmt.Printf(" Sum: %v, cost: %v \n", sum, cost)
		sum += cost
	}
	avg := sum / AVGTESTINGTIMES
	fmt.Printf(" Sum: %v, Avg: %v \n", sum, avg.Microseconds())
}

// Testing part 5
func IntTest() {
	fmt.Println("Int test")
	for i := 1000; i <= 10000; i += 1000 {
		IntCalCost(i)
		IntAvgCost(i)
	}
}

func StringTest() {
	fmt.Println("String test")
	for i := 1000; i <= 10000; i += 1000 {
		StringCalCost(i)
		StringAvgCost(i)
	}
}

func RationalNumTest() {
	fmt.Println("RationalNum test")
	for i := 1000; i <= 10000; i += 1000 {
		RationalNumCalCost(i)
		RationalNumAvgCost(i)
	}
}

func main() {
	//testing rational number
	RationalType.TestRationalNum()

	// 10 list and avg time test
	IntTest()
	StringTest()
	RationalNumTest()

	///////////////////////////////////////////////////////////////////
	// start := time.Now()
	// time.Sleep(time.Second * 1)
	// cost := time.Since(start)
	// fmt.Println(cost)
	// fmt.Println(cost.Microseconds())
}

// go run LinearInsertionSort.go
